### PR TITLE
working parser.mly

### PR DESCRIPTION
**Pull request by Contentmaudlin**

### CODE REVIEW

**Comment by j-hui** [comment]
@Contentmaudlin shouldn't we be mergin this into `frontend` instead of `master`? just wondering

File: src/frontend/parser.mly [hunk 1]
```diff
@@ -23,252 +23,256 @@
...
 ret_expr:
-| expr
-| (* good old bucket of nothing *)
-(* warn: this may produce S/R conflicts *)
+ expr									{ $1 } 
+| /* nothing */ 						{ Return Noexpr }
 
 expr:
-| asn_expr
+  asn_expr								{ [] }
```
Comment (j-hui) [inline src/frontend/parser.mly]: aww i like the extra `|`, it makes it look not so ugly

File: src/frontend/parser.mly [hunk 2]
```diff
@@ -23,252 +23,256 @@
...
+| /* nothing */ 						{ Return Noexpr }
 
 expr:
-| asn_expr
+  asn_expr								{ [] } 
 
 asn_expr:
-| unary_expr asn_op asn_expr
-| conditional_expr
+  unary_expr ASSIGN asn_expr			{ [] }
```
Comment (j-hui) [inline src/frontend/parser.mly]: we need to support `+=` and `*=` etc., i might have forgotten the following sort of rule:
```
asn_op:
| ASSIGN
| PLUS_ASN
| MINUS_ASN
| MULT_ASN
| DIV_ASN
| MOD_ASN
```

File: src/frontend/parser.mly [hunk 3]
```diff
@@ -23,252 +23,256 @@
...
 add_op:
-| PLUS
-| MINUS
+  PLUS									{ [] }
+| MINUS									{ [] }
 
 mult_op:
-| TIMES
-| DIVIDE
+  TIMES									{ [] }
```
Comment (j-hui) [inline src/frontend/parser.mly]: I think we need to put `MOD` – that is, `%` – over here too. i totally forgot about this but it's a damn useful operation

File: src/frontend/parser.mly [hunk 4]
```diff
@@ -23,252 +23,256 @@
...
 
 ns_decl:
-| NS ID ASSIGN LBRACE program RBRACE    { }
+  NS ID ASSIGN LBRACE program RBRACE    { }
 
-(* need to consider structs *)
+/* need to consider structs */
 let_decl:
-| LET type ID ASSIGN expr SEMICOLON { }
+  LET ID ASSIGN expr SEMI { }
```
Comment (j-hui) [inline src/frontend/parser.mly]: we called it `SEMI`? just double check to make sure that it's consistent with what we have in the scanner

File: src/frontend/parser.mly (hunk 2 above)
Comment (j-hui) [reply src/frontend/parser.mly]: actually we don't really _have_ to suppor this tbh

**Review by j-hui — CHANGES_REQUESTED** [review]
mostly look at `%` operator and assignment operators (we don't strictly have to support them, they shouldn't be used very often)

Other than that LGTM

File: src/frontend/parser.mly (hunk 1 above)
Comment (Contentmaudlin) [reply src/frontend/parser.mly]: Nah

File: src/frontend/parser.mly (hunk 2 above)
Comment (Contentmaudlin) [reply src/frontend/parser.mly]: Let me put this in

File: src/frontend/parser.mly (hunk 4 above)
Comment (Contentmaudlin) [reply src/frontend/parser.mly]: Yep!

**Pull request closed without merging**